### DISCUSSION TITLE
Roland KD-120 kick velocity

### DISCUSSION THREAD

**Discussion by thijstriemstra** [corrados/edrumulus#29] [General]
I'm testing the Roland KD-120 kick with an ESP32 and this kick config:

| Pad | Type | MIDI note | GPIO |
| --- | --- | --- | --- |
| Roland KD-120BK | Kick | 36 | 33 |

and it seems the velocity doesn't go higher than 80 or so.

Here's a video and below that a screenshot with the associated Ableton midi notes, where you can see the velocities. Those last 2 notes should be 127 IMO. I don't want to destroy this kick I borrowed and try to hit it even louder ;)

https://user-images.githubusercontent.com/305679/147893646-4d254d2f-3d7e-4de0-8e7b-9a56a521b39f.mp4

![kd120-velocity](https://user-images.githubusercontent.com/305679/147893553-f4ae594e-f7f6-4887-9329-dad32a29f07a.jpg)

**Comment by corrados** [comment]
You must adjust the Edrumulus pad parameters to get correct results. You should use the tools/edrumuluscontrol.m in Octave:
![grafik](https://user-images.githubusercontent.com/46655886/147950327-23fbb946-03e1-4c35-b8bd-6cc0690fbdb7.png)

1. Select the correct pad by using Pad Select or enable Auto Pad Select and hit your pad
2. Make sure the correct Pad Type is selected
3. Adjust the Threshold so that you do not hear unwanted MIDI notes. For the kick I would set the threshold pretty high.
4. Adjust the Sensitivity so that the hardest strike leads to a MIDI velocity of 127.
5. For your snare pad: Adjust the Pos Threshold so that MIDI control 16 starts to increase at the position at the mesh head where you want to get a different sound.
6. Adjust the Pos Sensitivity so that you get a MIDI control 16 value of 127 if you hit the pad at the most outer edge.
7. Adjust the RimShot Threshold until you do not get rim shot sounds when you only strike on the mesh had.

Write down the parameters since they are not stored in the Edrumulus software. You could hard code these parameters in edrumulus.ino in the section "// my prototype setup configuration..."

**Reply by thijstriemstra** [reply]
thanks for the feedback @corrados, I'll give that a try now. Should I merge main branch into thijstriemstra_prototype? Because it contains some rim shot fix(es) it looks like.

**Reply by thijstriemstra** [reply]
@corrados I'm probably doing something wrong because adjusting Threshold or Sensitivity doesn't seem to change anything.

- Started HairlessMIDI and loopMIDI
- Started Ableton and configured VST to playback Edrumulus MIDI notes
- Start octave, open the `tools/edrumuluscontrol.m` in Octave
- Press 'Save file and Run'

And set change pad to PD120. This is weird though, the snare uses PD120 as well, as seen here: https://github.com/corrados/edrumulus/blob/thijstriemstra_prototype/edrumulus.ino#L88 Maybe this is the issue in my setup?

![octave-test](https://user-images.githubusercontent.com/305679/147967035-efea7b73-fff2-489d-9193-acd90ff484a4.jpg)


Recorded midi:

![kd120-velocity2](https://user-images.githubusercontent.com/305679/147967151-ca3328c0-3fb0-4a13-a467-704a95b8da6d.jpg)

**Reply by thijstriemstra** [reply]
Also, why are both PD-5 tom pads set to different Edrumulus constants (CY-6 vs PD8), shouldn't they be the same; was this intentional?

https://github.com/corrados/edrumulus/blob/thijstriemstra_prototype/edrumulus.ino#L93

**Reply by corrados** [reply]
> Should I merge main branch into thijstriemstra_prototype? Because it contains some rim shot fix(es) it looks like.

You can do that. That should not hurt.

> I'm probably doing something wrong because adjusting Threshold or Sensitivity doesn't seem to change anything.
> And set change pad to PD120. This is weird though, the snare uses PD120 as well, as seen here: https://github.com/corrados/edrumulus/blob/thijstriemstra_prototype/edrumulus.ino#L88 Maybe this is the issue in my setup?

Good question... Have you tried the Auto Pad Select and have you hit the kick and snare and see if the settings change? Maybe that gives you a hint what could go wrong. Maybe you have adjusted the Threshold of a different pad input and not the one you actually want to change.

> Also, why are both PD-5 tom pads set to different Edrumulus constants (CY-6 vs PD8), shouldn't they be the same; was this intentional?

Because in the section "prototype setup configuration", I configured my setup for the Teensy which has the crash pad on input 4. For your prototype, you should change the set_pad_type          ( 4, Edrumulus::CY6 );  to set_pad_type          ( 4, Edrumulus::PD8 );

**Reply by thijstriemstra** [reply]
> You can do that. That should not hurt.

Did this in #32 

> For your prototype, you should change the set_pad_type ( 4, Edrumulus::CY6 ); to set_pad_type ( 4, Edrumulus::PD8 );

Changed this also in #32 

>  Have you tried the Auto Pad Select and have you hit the kick and snare and see if the settings change?

Nothing changes in the UI.

What I dont understand, does Octave allow me to tweak edrumulus on the esp32 in realtime? The midi inputs in the user interface suggest it can. Or should I manually update the .ino file and upload everytime?

I tried changing this for the kick, and then uploading the updated ino to the esp32:

```cpp
// see https://github.com/corrados/edrumulus/discussions/29
edrumulus.set_velocity_threshold(1, 28);
edrumulus.set_velocity_sensitivity(1, 25);
```

but this did not improve the max velocity, it only made it much less responsive.

**Reply by corrados** [reply]
Edrumuluscontrol.m does it in real-time. No need to flash the ESP32 for that.

**Reply by corrados** [reply]
Have you configured the return path in HairlessMIDI? You need to set both combo boxes to the correct value.

**Reply by thijstriemstra** [reply]
> Have you configured the return path in HairlessMIDI? You need to set both combo boxes to the correct value.

What should the return path be? I'm kinda confused by this setup..

> Edrumuluscontrol.m does it in real-time. 

I don't understand how octave controls the esp32, when the esp32 is connected through hairless midi.. some guidance would be really welcome here.

**Reply by thijstriemstra** [reply]
I now have following settings for loopMIDI:

![loop-midi-for-octave](https://user-images.githubusercontent.com/305679/148111560-6bd66237-f5d8-4d20-95cc-c698c53a67c5.jpg)

Configured Hairless MIDI IN to Edrumulus and MIDI OUT to Octave:

![hairless-octave](https://user-images.githubusercontent.com/305679/148111736-46b1fa74-e359-4a37-bb6c-9e922bbaea4e.jpg)

And Addictive Drums 2 has MIDI IN set to Octave:


![ad2-octave](https://user-images.githubusercontent.com/305679/148111925-f509e193-e77f-4f2a-888f-0ade57bda131.jpg)

In Octave, when moving the threshold to 5 makes it send out spurious MIDI notes, which makes me believable this is a working test setup, correct @corrados? Unfortunately moving treshold and sensitivity on pad 1 (kick) doesn't do anything to the velocity.

![octave-test2](https://user-images.githubusercontent.com/305679/148112136-f324810b-9575-482f-ae10-cffde234dc46.jpg)

**Reply by thijstriemstra** [reply]
>  Unfortunately moving treshold and sensitivity on pad 1 (kick) doesn't do anything to the velocity.

Spoke to soon... i have now spurious kicks that can be adjusted with treshold.. and sensitivity makes them louder (yay!)

**Reply by thijstriemstra** [reply]
So I ended up with these settings for the KD-120:

```c++
  // see https://github.com/corrados/edrumulus/discussions/29
  edrumulus.set_velocity_threshold(1, 9);     // default is 8
  edrumulus.set_velocity_sensitivity(1, 19);  // default is 9
```

I suppose at some point there will be a new pad type in edrumulus? Should I make a pull request?

**Reply by corrados** [reply]
> Should I make a pull request?

You mean on your prototype branch? Why not? :-)

Would be interesting to see if the positional sensing works on your Drum-tec snare pad.

**Reply by thijstriemstra** [reply]
> Would be interesting to see if the positional sensing works on your Drum-tec snare pad.

See #30

**Reply by thijstriemstra** [reply]
> So I ended up with these settings for the KD-120:

Here's a video of the kick. With tuning the parameters I lost some sensitivity in order to gain velocity but that's fine. You can see the softest kicks are not triggered, but I couldn't get to 127 at a hard kick otherwise, if that makes sense.

https://user-images.githubusercontent.com/305679/148127302-fbc36fea-760d-4637-8096-5172efb038aa.mp4

![kd120-velocity3](https://user-images.githubusercontent.com/305679/148127804-ecdc4e52-8e96-42ad-80fd-8cf067f74578.jpg)

**Reply by corrados** [reply]
Your MIDI recording looks like you are clipping the MIDI notes too early at 127. I would reduce the sensitivity a bit.

It is not clear to me that you have to raise the threshold to have enough range for the sensitivity setting. If you set the threshold to low values, do you then set the sensitivity to 31 and still do not reach the MIDI note 127?

**Reply by thijstriemstra** [reply]
I will get back to you on this.

Btw, is it useful if I do a .wav recording of the kick (so it can be hopefully included in edrumulus as a pad type)?

**Reply by corrados** [reply]
Yes, a recording would be good to have. The wave should contain very low level/mid level and very loud hits.

**Reply by thijstriemstra** [reply]
> The wave should contain very low level/mid level and very loud hits.

Here's a zip with (mono) recordings of the kick. If you need stereo, let me know.

[roland-kd-120-kick.zip](https://github.com/corrados/edrumulus/files/7833636/roland-kd-120-kick.zip)

**Reply by corrados** [reply]
I added your signal to the Git repo: https://github.com/corrados/edrumulus/commit/e7a1616d4bf01c465aad7eaf07aef1fe027a796b. Thank you for the recording.

**Comment by thijstriemstra** [comment]
> What I dont understand, does Octave allow me to tweak edrumulus on the esp32 in realtime? 

I wrote this a while back, but I think I have a better understanding of edrumulus now. Basically, every parameter/edrumulus setting can be changed by sending one or more MIDI commands (using serial [only]?), eg. a program written with Octave.

Having a simplified, well documented, structure of the edrumulus architecture in the readme **and** docs would be really useful.

If it's all midi, what midi notes etc are we talking about?

**Reply by corrados** [reply]
> Having a simplified, well documented, structure of the edrumulus architecture in the readme and docs would be really useful.

Recently, I started to write a document: https://github.com/corrados/edrumulus/blob/main/doc/manual.md. But I'll concentrate on other things right now. Feel free to expand that document and create a pull request.

**Reply by corrados** [reply]
> If it's all midi, what midi notes etc are we talking about?

It's all in the edrumulus.ino file.
